### PR TITLE
Draft: Support regular spring JDBC transactions in the Exposed `SpringTransactionManager`

### PR DESCRIPTION
**Pull request by bystam**
#### Description

The Exposed `SpringTransactionManager` currently does not make Spring JDBC constructs take part in the transaction. In short:

```kotlin
@Transactional
fun insertTwoThings() {
    // these two statements will NOT be run in a transaction
    jdbcTemplate.executeUpdate("INSERT INTO foo VALUES ('bar1')")
    jdbcTemplate.executeUpdate("INSERT INTO foo VALUES ('bar2')")
}
```

This draft tries to start supporting this by adding functionality to `SpringTransactionManager` borrowed from the basic `DataSourceTransactionManager`.

**Detailed description**:
Caveat: I am not an expert on neither Exposed or Spring JDBC. This is my best effort attempt at simply reading source code.

- In short, Spring JDBC manages its transactions by pinning `ConnectionHolder` instances to thread locals inside the `TransactionSynchronizationManager`. Spring JDBC classes like `JdbcTemplate` then reuse those `ConnectionHolder` instances to make sure a single autocommit-disabled connection is shared throughout the transaction
- Since the Exposed `SpringTransactionManager` does not interact with `TransactionSynchronizationManager` at all - it does not influence transaction semantics in Spring JDBC.
- This PR makes a quick attempt at blending `SpringTransactionManager` and `DataSourceTransactionManager` in order to make `@Transactional` influence both Exposed and Spring JDBC simultaneously.

**But why?**
Since Exposed depends on `spring-jdbc`, in my mind it makes sense that it would honor the transaction semantics of its dependency. In comparison, when using JPA/Hibernate in Spring, the `JpaTransactionManager` introduces JPA-specific transaction behaviour but ALSO make sure to adhere to the expected semantics of 

**Why a draft?**
This is written by basically cherry-picking `DataSourceTransactionManager` code into `SpringTransactionManager`. I feel like I managed to get a pretty decent high level understanding of how Spring JDBC manages transactions, but there's plenty of low level stuff in there I am uncertain about whether it made sense to bring or not.

`DataSourceTransactionManager.doBeing(...)` had quite a lot of code that basically followed the pattern "remember what settings the connection had before so that we can restore it". I am not entirely certain all of that belongs in `SpringTransactionManager` or not if it will compete with Exposed constructs.

---

#### Type of Change

Please mark the relevant options with an "X":
- [x] Bug fix
- [x] New feature
- [ ] Documentation update

Updates/remove existing public API methods:
- [ ] Is breaking change

Affected databases:
- [ ] MariaDB
- [ ] Mysql5
- [ ] Mysql8
- [ ] Oracle
- [ ] Postgres
- [ ] SqlServer
- [ ] H2
- [ ] SQLite

#### Checklist

- [x] Unit tests are in place
- [ ] The build is green (including the Detekt check)
- [ ] All public methods affected by my PR has up to date API docs
- [ ] Documentation for my change is up to date

---

#### Related Issues
https://youtrack.jetbrains.com/issue/EXPOSED-657/Transaction-manager-provided-by-Exposed-is-not-compatible-with-JdbcTemplate

### CODE REVIEW

**Comment by bystam** [comment]
Apologize for the failing PR title/commit message/detekt checks. I will make sure to get that sorted.

File: spring-transaction/src/main/kotlin/org/jetbrains/exposed/spring/SpringTransactionManager.kt [hunk 1]
```diff
@@ -63,16 +69,23 @@ class SpringTransactionManager(
             manager = threadLocalTransactionManager,
             outerManager = outerManager,
             outerTransaction = outer,
-        )
+        ).apply {
+            setConnectionHolder(
+                TransactionSynchronizationManager.getResource(dataSource) as? ConnectionHolder
+            )
```
Comment (bystam) [inline spring-transaction/src/main/kotlin/org/jetbrains/exposed/spring/SpringTransactionManager.kt:75]: Detect any existing `ConnectionHolder` set as a thread-local and reuse it

File: spring-transaction/src/main/kotlin/org/jetbrains/exposed/spring/SpringTransactionManager.kt [hunk 2]
```diff
@@ -115,6 +130,50 @@ class SpringTransactionManager(
...
+
+            // Bind the connection holder to the thread.
+            if (trxObject.isNewConnectionHolder){
+                TransactionSynchronizationManager.bindResource(dataSource, trxObject.getConnectionHolder())
+            }
+        } catch (ex: Throwable) {
+            DataSourceUtils.releaseConnection(con, dataSource)
+            trxObject.setConnectionHolder(null)
+            throw CannotCreateTransactionException("Could not open JDBC Connection for transaction", ex)
+        }
```
Comment (bystam) [inline spring-transaction/src/main/kotlin/org/jetbrains/exposed/spring/SpringTransactionManager.kt:176]: This block is more or less lifted from `DataSourceTransactionManager`:

https://github.com/spring-projects/spring-framework/blob/main/spring-jdbc/src/main/java/org/springframework/jdbc/datasource/DataSourceTransactionManager.java#L261

Comment (bystam) [reply spring-transaction/src/main/kotlin/org/jetbrains/exposed/spring/SpringTransactionManager.kt:176]: I intend to also check the `JpaTransactionManager`, which is an implementation in the same seat as the Exposed one, to see if there is anything to deduce from its similarities/differences to `DataSourceTransactionManager`

https://github.com/spring-projects/spring-framework/blob/main/spring-orm/src/main/java/org/springframework/orm/jpa/JpaTransactionManager.java

File: spring-transaction/src/main/kotlin/org/jetbrains/exposed/spring/SpringTransactionManager.kt [hunk 3]
```diff
@@ -137,6 +196,23 @@ class SpringTransactionManager(
...
+
+        // Reset connection.
+        val con: Connection = trxObject.getConnectionHolder().getConnection()
+        try {
+            if (trxObject.mustRestoreAutoCommit) {
+                con.autoCommit = true
+            }
+            DataSourceUtils.resetConnectionAfterTransaction(con, trxObject.previousIsolationLevel, trxObject.isReadOnly)
+        } catch (ex: Throwable) {
+            logger.debug("Could not reset JDBC Connection after transaction", ex)
```
Comment (bystam) [inline spring-transaction/src/main/kotlin/org/jetbrains/exposed/spring/SpringTransactionManager.kt:214]: Some logging came with and ended up identical to the Spring JDBC ones. Obviously those should be cleaned up.

File: spring-transaction/src/main/kotlin/org/jetbrains/exposed/spring/SpringTransactionManager.kt [hunk 4]
```diff
@@ -212,7 +290,15 @@ class SpringTransactionManager(
         override fun isRollbackOnly() = isRollback
 
         override fun flush() {
-            // Do noting
+            TransactionSynchronizationUtils.triggerFlush()
```
Comment (bystam) [inline spring-transaction/src/main/kotlin/org/jetbrains/exposed/spring/SpringTransactionManager.kt:293]: `DataSourceTransactionManager.DataSourceTransactionObject` (which is the transaction object) had this on `flush`. I don't fully understand why it's there, yet, sadly.

File: spring-transaction/src/main/kotlin/org/jetbrains/exposed/spring/SpringTransactionManager.kt [hunk 5]
```diff
@@ -212,7 +290,15 @@ class SpringTransactionManager(
...
+        }
+    }
+
+    class ExposedConnectionHandle(
+        val transaction: Transaction
+    ) : ConnectionHandle {
+        override fun getConnection(): Connection {
+            return transaction.connection.connection as Connection
```
Comment (bystam) [inline spring-transaction/src/main/kotlin/org/jetbrains/exposed/spring/SpringTransactionManager.kt:301]: This class is actually more or less the thing that makes the whole implementation work.

For Exposed and Spring JDBC to co-exist, they need to both see the same connection. Spring JDBC fetches its connection by calling something like:
```java
ConnectionHolder holder = (ConnectionHolder) TransactionSynchronizationManager.getResource(dataSource);
holder.getConnection()
```

so for them to share it, the Exposed transaction manager needs to install a `ConnectionHolder` that provides the same connection using `TransactionSynchronizationManager.bindResource(dataSource, <some connection holder>)`. The `ConnectionHolder` class of spring can take a "pointer to a connection" called a `ConnectionHandle`, which is what I implement here.

This is done upstairs like
```kotlin
trxObject.setConnectionHolder(
    ConnectionHolder(ExposedConnectionHandle(transaction))
)
trxObject.isNewConnectionHolder = true
...

if (trxObject.isNewConnectionHolder){
    TransactionSynchronizationManager.bindResource(dataSource, trxObject.getConnectionHolder())
}
```

File: spring-transaction/src/main/kotlin/org/jetbrains/exposed/spring/SpringTransactionManager.kt [hunk 6]
```diff
@@ -115,6 +130,50 @@ class SpringTransactionManager(
...
         }
+
+        lateinit var con: Connection
+        try {
+            if (!trxObject.hasConnectionHolder() ||
+                trxObject.getConnectionHolder().isSynchronizedWithTransaction
+            ) {
+                trxObject.setConnectionHolder(
+                    ConnectionHolder(ExposedConnectionHandle(transaction))
+                )
```
Comment (bystam) [inline spring-transaction/src/main/kotlin/org/jetbrains/exposed/spring/SpringTransactionManager.kt:141]: This `ExposedConnectionHandle` is the thing that makes this all work. See comment downstairs.

**Comment by bystam** [comment]
I obviously broke a bunch of other tests using this implementation 🙃 - looking into it

**Comment by bystam** [comment]
Okay - my attempt broke quite a lot of tests it looks like. Might have been a bit naive to simply blend them like I did. Maybe things need to be integrated more deeply into the Exposed `TransactionManager` rather than handled on the outside.

**Comment by FullOfOrange** [comment]
When I first proposed to re-implement the Spring transaction manager, I thought about implementing an Exposed TransactionManager for Spring, but since it is hard work, I thought it would be more efficient to integrate ThreadLocalTransactionManager's transaction management with Spring's transaction lifecycle. 

I used ThreadLocalTransactionManager to integrate with Spring, but it seemed a bit too much to fulfill the basic requirements of Spring, such as the SQLException not rollback issue. On second thought, it would be nice to rewrite the TransactionManager implementation, even if it's difficult.

**Comment by bystam** [comment]
Looking more closely at the failing tests - I got the sense that all the "extra stuff" that I cherry-picked from the `DataSourceTransactionManager` that manipulated the connection might have been what made it not work.

I decided to scratch most of that, since the regular Exposed `TransactionManager` handles a bunch of connection configuration itself. 

What's left is simply the `TransactionSynchronizationManager` + `ConnectionHolder` dance - which I believe was the core part to make this actually work. Now the tests seem green to me!

The only bit I am curious about is whether it's important to set the `isSynchronizedWithTransaction` flag or not. If I use it exactly like `DataSourceTransactionManager` does then I get some failures, but that's because `useSavepointForNestedTransaction` is set to false. I don't totally know how to write a test that might reproduce any failures with misusing that flag - or if it even matters.

**Comment by bystam** [comment]
> When I first proposed to re-implement the `SpringTransactionManager`, I thought about implementing an Exposed `TransactionManager` for Spring, but since it is hard work, I thought it would be more efficient to integrate ThreadLocalTransactionManager's transaction management with Spring's transaction lifecycle. (I thought the SpringTransactionManager would just set @transactional as a replacement for transaction {})
> 
> I used ThreadLocalTransactionManager to integrate with Spring, but it seemed a bit too much to fulfill the basic requirements of Spring, such as the SQLException not rollback issue. On second thought, it would be nice to rewrite the TransactionManager implementation, even if it's difficult.

I see! I think I managed to get it to work now actually. Though I am not entirely sure how I would write a test that might catch any nasty edge cases. All the existing tests and the new ones I added are green, though!

**Comment by bystam** [comment]
I created a copy of the existing `ExposedTransactionManagerTest` that runs basically the same tests but using springs `JdbcClient` to verify all the same cases.

Some of them behaved differently, but that is intentional. Exposed requires a transaction in order to operate, but `JdbcClient` does not - and instead runs in autocommit mode if none is present.

File: spring-transaction/src/test/kotlin/org/jetbrains/exposed/spring/TransactionSynchronizationTest.kt [hunk 7]
```diff
@@ -0,0 +1,96 @@
+package org.jetbrains.exposed.spring
+
+import junit.framework.TestCase.assertTrue
+import org.junit.Before
+import org.springframework.transaction.TransactionDefinition
+import org.springframework.transaction.support.TransactionSynchronization
+import org.springframework.transaction.support.TransactionSynchronizationManager
+import kotlin.test.Test
+
+class TransactionSynchronizationTest : SpringTransactionTestBase() {
```
Comment (bystam) [inline spring-transaction/src/test/kotlin/org/jetbrains/exposed/spring/TransactionSynchronizationTest.kt:10]: This behaviour is somewhat tricky. The full `TransactionSynchronization` behaviour is actually not supported right now - because we handle savepoints outside of Springs knowledge, which means that the savepoint callbacks specifically do not trigger.

While that might be a somewhat incomplete support - I personally feel that covering 95% of the common use cases is better than not covering them, as listening to savepoint callbacks should be a quite niche thing.

**Comment by joc-a** [comment]
Hi @bystam, thank you for your contribution. Please rebase on the `main` branch to make the Pull Request Title Validation check pass.

File: spring-transaction/src/main/kotlin/org/jetbrains/exposed/spring/SpringTransactionManager.kt (hunk 4 above)
Comment (bystam) [reply spring-transaction/src/main/kotlin/org/jetbrains/exposed/spring/SpringTransactionManager.kt]: Reading up on it more closely - it looks like it was added to `DataSourceTransactionManager` to be compatible with a scenario where you use it combined with things like JPA but without using the `JpaTransactionManager`.

This makes it feel a bit niche to me - and I decided to restore the "Do nothing" case that was there before

https://github.com/spring-projects/spring-framework/issues/19413

**Comment by bystam** [comment]
What do you think about the state of this PR? Does it make sense to you to support this?

At our company we use Exposed, but we also use other things in certain services. For that purpose our shared libraries that interact with the database (such as postgres-backed queue libraries) has been written in regular Spring JDBC to usable everywhere.

But that's how I stumbled upon this behaviour in Exposed - because we are currently having race condition issues since our code doesn't run in transactions as expected.

So from our point of view this is much desired. :) but I'm open to feedback related to anything lacking here.

**Comment by bystam** [comment]
@bog-walk @joc-a apologies for the ping directly - but I am not entirely sure whether this is currently under consideration or not.

I of course know that I cannot demand anything to be merged :) but it would be important to us to know whether this is a sound direction or not for Exposed, because we encountered the problem explained on YouTrack in our own system and would like to know if we should re-consider our implementations where we use Exposed or if something like this could be merged to solve it.

**Comment by thijsnederlof** [comment]
Any update on this? Ghosting a contributor like that is not very nice.

**Comment by bystam** [comment]
This one was submitted from my forked `main` branch - I created a copy of this that instead comes from a feature branch as intended:

https://github.com/JetBrains/Exposed/pull/2675
